tool/tsh/exec_command_test.go

```diff
@@ -0,0 +1,221 @@
+		{
+			name: "check environment",
+			cmd: func(t *testing.T) *execCommand {
+				app := kingpin.New("my test app", "")
+				exec := newExecCommand(app)
+				exec.arguments = []string{"-c", `
+cd $0
+echo $TSH_COMMAND >> tsh_command.env
+echo $TSH >> tsh.env`, tmp}
+				exec.shell = "bash"
+				return exec
+			},
+			args: args{cf: &CLIConf{executablePath: "/bin/tsh"}},
+			postCheck: func(t *testing.T) {
+				checkFile := func(fn string, contents string) {
+					bytes, err := os.ReadFile(path.Join(tmp, fn))
+					require.NoError(t, err)
+					require.Equal(t, contents, string(bytes))
+				}
+
+				checkFile("tsh_command.env", "/bin/bash\n")
+				checkFile("tsh.env", "/bin/tsh\n")
+			},
+			wantErr: false,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			exec := tt.cmd(t)
+			err := exec.runCommand(tt.args.cf)
+
+			if tt.wantErr {
+				require.Error(t, err)
+			} else {
+				require.NoError(t, err)
+			}
+
+			if tt.postCheck != nil {
+				tt.postCheck(t)
+			}
+		})
+	}
+}
+
+func Test_findShell(t *testing.T) {
+	mkshell := func(fn string) {
+		// use existing `sh` binary for tests.
+		shPath, err := os_exec.LookPath("sh")
+		require.NoError(t, err)
+
+		file, err := os.ReadFile(shPath)
+		require.NoError(t, err)
+
+		err = os.WriteFile(fn, file, 0777)
+		require.NoError(t, err)
+	}
+
+	tmp := t.TempDir()
+
+	tests := []struct {
+		name string
+		init func(t *testing.T) string
+		want string
+	}{
+		{
+			name: "return custom shell",
+			init: func(t *testing.T) string {
+				err := os.MkdirAll(path.Join(tmp, "custom"), 0777)
+				require.NoError(t, err)
+
+				dummyShell := path.Join(tmp, "custom", "dummysh")
+				mkshell(dummyShell)
+
+				return dummyShell
+			},
+			want: path.Join(tmp, "custom", "dummysh"),
+		},
+
+		{
+			name: "return $SHELL",
+			init: func(t *testing.T) string {
+				err := os.MkdirAll(path.Join(tmp, "SHELL"), 0777)
+				require.NoError(t, err)
+
+				dummyShell := path.Join(tmp, "SHELL", "dummysh")
+				mkshell(dummyShell)
+
+				t.Setenv("SHELL", dummyShell)
+
+				return ""
+			},
+			want: path.Join(tmp, "SHELL", "dummysh"),
+		},
+
+		{
+			name: "return sh from PATH",
+			init: func(t *testing.T) string {
+				err := os.MkdirAll(path.Join(tmp, "PATH_SH"), 0777)
+				require.NoError(t, err)
+
+				dummyShell := path.Join(tmp, "PATH_SH", "sh")
+				mkshell(dummyShell)
+
+				// unset shell, set path
+				t.Setenv("SHELL", "")
+				t.Setenv("PATH", path.Join(tmp, "PATH_SH"))
+
+				return ""
+			},
+			want: path.Join(tmp, "PATH_SH", "sh"),
+		},
+
+		{
+			name: "return bash from PATH",
+			init: func(t *testing.T) string {
+				err := os.MkdirAll(path.Join(tmp, "PATH_BASH"), 0777)
+				require.NoError(t, err)
+
+				dummyShell := path.Join(tmp, "PATH_BASH", "bash")
+				mkshell(dummyShell)
+
+				// unset shell, set path
+				t.Setenv("SHELL", "")
+				t.Setenv("PATH", path.Join(tmp, "PATH_BASH"))
+
+				return ""
+			},
+			want: path.Join(tmp, "PATH_BASH", "bash"),
+		},
```

Comment:
These three tests kind of test the behavior of `exec.LookPath`, aren't they? Wouldn't it be better to assert that `exec.LookPath` was called with correct arguments? I'm totally new to writing tests in Go and statically compiled langs in general, so just let me know if it wouldn't be feasible. ;) I know that it's not as easy to use mocks/spies there.

I'd also appreciate any resources on that topic!

Reply:
I think the typical approach here is an object-local private field that gets overridden for tests. For stateless functions, whey often consumes an argument that is different between production code and unit tests. The "injection point" is typically a callback function or, for when more complex behavior is needed, an interface. 

Example of function callback as part of object: 

https://github.com/gravitational/teleport/blob/8c46b21e2655ef646636f54989c93130471a83dc/lib/auth/oidc_test.go#L231-L241

Example of callback parameter to function: 

https://github.com/gravitational/teleport/blob/master/tool/tsh/tsh_test.go#L784-L786

Example of interface field for testing purposes:

https://github.com/gravitational/teleport/blob/master/tool/tsh/tsh.go#L284-L285

Hope that helps, but don't hesitate to ping me on Slack if needed 😄

Reply:
Ah, forgot to comment the test itself: I am *not* testing the `exec.LookPath` here, I really want to test the actual execution. I could mock `exec.LookPath`, but I don't think this is necessary.

Reply:
@ravicious done:  https://github.com/gravitational/teleport/pull/12629/commits/14cabbb62a75c24ee2d9f8a0e8fed937029987c4